### PR TITLE
[FLINK-13782][table-api] Implement type strategies for comparison expressions

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

This PR implements both input and output type strategies for comparison expressions.

## Verifying this change
All current expression tests should pass

Added test:
* org.apache.flink.table.types.inference.ComparableInputTypeStrategyTests
* TypeStrategiesTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/ComparableInputTypeStrategyTests.java [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+					distinctType("type", DataTypes.INT()).nullable()
+				),
+
+			TestSpec
+				.forStrategy(
+					"Everything is comparable with null type",
+					InputTypeStrategies.TWO_EQUALS_COMPARABLE)
+				.calledWithArgumentTypes(
+					DataTypes.INT(),
+					new AtomicDataType(new NullType())
```
Comment (twalthr) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/ComparableInputTypeStrategyTests.java:198]: Use DataTypes.NULL

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java [hunk 2]
```diff
@@ -58,15 +42,16 @@
...
 import static org.apache.flink.table.types.inference.InputTypeStrategies.sequence;
 import static org.apache.flink.table.types.inference.InputTypeStrategies.varyingSequence;
-import static org.apache.flink.util.CoreMatchers.containsCause;
-import static org.hamcrest.CoreMatchers.equalTo;
-import static org.junit.Assert.assertThat;
 
 /**
  * Tests for built-in {@link InputTypeStrategies}.
  */
 @RunWith(Parameterized.class)
```
Comment (twalthr) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java:49]: We can move this annotation to the test base.

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java [hunk 3]
```diff
@@ -58,15 +42,16 @@
...
-public class InputTypeStrategiesTest {
+public class InputTypeStrategiesTest extends InputTypeStrategiesTestBase {
+
+	public InputTypeStrategiesTest(TestSpec testSpec) {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java:52]: why is this required?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ComparableTypeStrategy.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+				secondLogicalType instanceof LegacyTypeInformationType) {
+			return true;
+		}
+
+		// everything is comparable with null, it should return null in that case
+		if (hasRoot(firstLogicalType, LogicalTypeRoot.NULL) || hasRoot(secondLogicalType, LogicalTypeRoot.NULL)) {
+			return true;
+		}
+
+		if (hasRoot(firstLogicalType, LogicalTypeRoot.STRUCTURED_TYPE) &&
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ComparableTypeStrategy.java:112]: nit: check for same type root and then use a switch/case in a separate method?

Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ComparableTypeStrategy.java:112]: Very good comment! It simplified the code a lot!

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java (hunk 2 above)
Comment (dawidwys) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java:49]: I thought the annotation cannot be on the abstract class. Or to say it differently that the `Parameter` can be in a separate class than `Parameters`.

Changed that.

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java (hunk 3 above)
Comment (dawidwys) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java:52]: As above. Did not know the `Parameter` can be in a separate class than `Parameters`. Therefore I was passing the parameters through constructor. As I moved the `RunWith` to the base class, it is no longer needed.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ComparableTypeStrategy.java [hunk 5]
```diff
@@ -97,100 +98,84 @@ public ArgumentCount getArgumentCount() {
...
 
-	private boolean areDistinctTypesComparable(LogicalType firstLogicalType, LogicalType secondLogicalType) {
-		DistinctType firstDistinctType = (DistinctType) firstLogicalType;
-		DistinctType secondDistinctType = (DistinctType) secondLogicalType;
-		return firstLogicalType.copy(true).equals(secondLogicalType.copy(true)) &&
+	private boolean areDistinctTypesComparable(LogicalType firstType, LogicalType secondType) {
+		DistinctType firstDistinctType = (DistinctType) firstType;
+		DistinctType secondDistinctType = (DistinctType) secondType;
+		return firstType.equals(secondType) &&
 			areComparable(firstDistinctType.getSourceType(), secondDistinctType.getSourceType());
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ComparableTypeStrategy.java:168]: @dawidwys One last minute comment: We need to remove the nullability for all calls of `areComparable` here and for other types. Maybe it is better to introduce an intermediate function call.
